crates/red_knot_python_semantic/src/types.rs

```diff
@@ -393,6 +400,69 @@ impl<'db> Type<'db> {
+        IntersectionBuilder::build_truthy(db, *self)
+    }
+
+    /// Return the `Falsy` subset of this type (intersection with Falsy).
+    #[must_use]
+    pub fn falsy(&self, db: &'db dyn Db) -> Type<'db> {
+        self.falsy_set(db)
+            .unwrap_or_else(|| IntersectionBuilder::build_falsy(db, *self))
+    }
+
+    /// Returns, if it can be expressed, the set of values that are falsy in this type. This is
+    /// defined for some builtin types (e.g. int, str, ...)
+    #[must_use]
+    pub fn falsy_set(&self, db: &'db dyn Db) -> Option<Type<'db>> {
+        match self {
+            Self::Instance(class) => match class.known(db) {
+                // For the following types, we know the logic for falsiness and can represent
+                // specific instances of that type, so we can give back a precise subset.
+                Some(KnownClass::Bool) => Some(Self::BooleanLiteral(false)),
+                Some(KnownClass::Int) => Some(Self::IntLiteral(0)),
+                Some(KnownClass::Bytes) => Some(Self::BytesLiteral(BytesLiteralType::empty(db))),
+                Some(KnownClass::Str) => Some(Self::StringLiteral(StringLiteralType::empty(db))),
+                Some(KnownClass::Tuple) => Some(Self::Tuple(TupleType::empty(db))),
+                _ => match self.bool(db) {
+                    // When a `__bool__` signature is `AlwaysFalse`, it means that the falsy set is
+                    // the whole type (that we had as input), and the truthy set is empty.
+                    Truthiness::AlwaysFalse => Some(*self),
+                    Truthiness::AlwaysTrue => Some(Type::Never),
+                    Truthiness::Ambiguous => None,
+                },
+            },
+            Type::LiteralString => Some(Self::StringLiteral(StringLiteralType::empty(db))),
+            _ => match self.bool(db) {
+                Truthiness::AlwaysFalse => Some(*self),
+                Truthiness::AlwaysTrue => Some(Type::Never),
+                Truthiness::Ambiguous => None,
+            },
+        }
+    }
+
+    /// Returns, if it can be expressed, the set of values that are truthy in this type. This is
+    /// rarely defined, with the exception of the `builtins.bool` type.
+    #[must_use]
+    pub fn truthy_set(&self, db: &'db dyn Db) -> Option<Type<'db>> {
+        match self {
+            Type::None => Some(Type::Never),
+            Type::Instance(class) => match class.known(db) {
+                Some(KnownClass::Bool) => Some(Type::BooleanLiteral(true)),
+                _ => match self.bool(db) {
+                    // When a `__bool__` signature is `AlwaysTrue`, it means that the truthy set is
+                    // the whole type (that we had as input), and the falsy set is empty.
+                    Truthiness::AlwaysTrue => Some(*self),
+                    Truthiness::AlwaysFalse => Some(Type::Never),
+                    Truthiness::Ambiguous => None,
+                },
+            },
+            _ => None,
```

Comment:
Similar to the `falsy_set` method above, I think this should have a fallback branch for arbitrary types where `self.bool()` returns either `AlwaysTruthy` or `AlwaysFalsy`. And if you add such a fallback branch, you should be able to get rid of the `Type::None` branch, since `Type::None` is known to be an `AlwaysFalsy` type.

```suggestion
            Type::Instance(class) => match class.known(db) {
                Some(KnownClass::Bool) => Some(Type::BooleanLiteral(true)),
                _ => match self.bool(db) {
                    // When a `__bool__` signature is `AlwaysTrue`, it means that the truthy set is
                    // the whole type (that we had as input), and the falsy set is empty.
                    Truthiness::AlwaysTrue => Some(*self),
                    Truthiness::AlwaysFalse => Some(Type::Never),
                    Truthiness::Ambiguous => None,
                },
            },
            _ => match self.bool(db) {
                Truthiness::AlwaysFalse => Some(*self),
                Truthiness::AlwaysTrue => Some(Type::Never),
                Truthiness::Ambiguous => None,
            },
```

You could possibly also add branches here for some of the types you enumerated in the `always_falsy` method above. E.g. `str & Truthy` could be simplified to `str & ~Literal[""]`: the set of all truthy strings is exactly equal to the set of all strings that are not the empty string. But I'm not sure if that's a good idea or not. _Is_ it a simplification? It's actually a more complicated way to express the same type, even if it's easier for me to understand.

Reply:
> You could possibly also add branches here for some of the types you enumerated in the always_falsy method above. E.g. str & Truthy could be simplified to str & ~Literal[""]: the set of all truthy strings is exactly equal to the set of all strings that are not the empty string. But I'm not sure if that's a good idea or not. Is it a simplification? It's actually a more complicated way to express the same type, even if it's easier for me to understand.

I think we might want to keep this idea somewhere, but this could either become a critical way to improve the logic of simplification or on the opposite it could complicate things.